schema/crdb/README.adoc

```diff
@@ -126,19 +126,31 @@ same `NEW_VERSION`:**, then your `OLD_VERSION` has changed and so _your_
    new version that came in from "main").
 * Update the version in `dbinit.sql` to match the new `NEW_VERSION`.
 
-=== General notes
+=== Constraints on Schema Updates
 
-CockroachDB's representation of the schema includes some opaque
-internally-generated fields that are order dependent, like the names of
-anonymous CHECK constraints.  Our schema comparison tools intentionally ignore
-these values. As a result, when performing schema changes, the order of new
-tables and constraints should generally not be important.
+==== Adding a new column without a default value [[add_column_constraint]]
+
+When adding a new column to an existing table, that column must
+contain a default to help back-fill existing rows in that table which may
+exist. Without this default value, the schema upgrade might fail with
+an error like `null value in column "..." violates not-null constraint`.
+Unfortnately, it's possible that the schema upgrade might NOT fail with that
+error, if no rows are present in the table when the schema is updated. This
+results in an inconsistent state, where the schema upgrade might succeed on
+some deployments but fail on others.
+
+If you'd like to add a column without a default value, we recommend
+doing the following, if a `DEFAULT` value makes sense for a one-time update:
 
-As convention, however, we recommend keeping the `db_metadata` file at the end
-of `dbinit.sql`, so that the database does not contain a version until it is
-fully populated.
+1. Adding the column with a `DEFAULT` value.
+2. Dropping the `DEFAULT` constraint.
 
-=== Scenario-specific gotchas
+If a `DEFAULT` value does not make sense, the following may be necessary
+instead:
+
+1. Add a column without a `NOT NULL` constraint.
+2. Ensure that all data has migrated to a non-nullable value, eventually.
+3. Once all data has been migrated to a non-nullable value, add a `NOT NULL` constraint.
```

Comment:
As a practical matter, can we ever be sure that all customer data has migrated this far?  I guess if we established a requirement that the cluster boot/run at schema level `N` before being allowed to migrate to `N+M`, it might be possible.  Which is a long way of saying that "If all data..." might be more apt than "Once all data...".

Reply:
Yeah, I was intentionally very handwavey with "how do you be certain the migration has completed", because it's very dependent on your data.

If it's possible to back-fill this data by observing the database, our *current* schema-updating system acts on a system with no other concurrent queries, so it should be possible to back-fill these values based on other tables and then make the conversion.

If backfilling this data requires reading data outside the database -- maybe we want to add a non-nullable field to inventory? -- then this is *much* harder. We basically rely on Nexus performing work *after* some schema updates are applied, but before the "nullability" constraint has been fully dropped. **Right now**, we don't really have a great way of doing this, and we'll probably need to build more mechanism to make this possible.

Reply:
I proposed more specific language.  I think the only realistic choice for the time being is to migrate the data yourself during the update.  We've done that a bunch of times.

An alternative might be to ship the update over two releases: one that adds the column and whose Nexus backfills data in the meantime; and one that makes the column required.  But this doesn't work unless there's some kind of interlock to ensure that the backfill process has finished before somebody starts the second upgrade.

An alternative is to ship an ad hoc Rust program alongside schema-updater (or a subcommand of that program) that backfills the data and maybe knows to do it between two specific schema versions.  This is probably the most promising but let's not do that if we don't have to.